File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/ReaderError.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.feature.scripture.R
import com.nexussoft.verbum.clients.BibleClientException
import com.nexussoft.verbum.models.BookId
import com.nexussoft.verbum.models.PassageReference

/** What the reader can tell the user when a chapter fails to load (docs/PRODUCT.md §52). */
sealed interface ReaderError {
    data class ChapterUnavailable(val reference: PassageReference) : ReaderError
    data class UnknownBook(val bookId: BookId) : ReaderError
    data object Offline : ReaderError
    data object Unexpected : ReaderError

    /** Title and message as string resources; the argument (if any) is filled in by the view. */
    val titleRes: Int
        get() = when (this) {
            is ChapterUnavailable -> R.string.err_chapter_unavailable_title
            is UnknownBook -> R.string.err_unknown_book_title
            Offline -> R.string.err_offline_title
            Unexpected -> R.string.err_unexpected_title
        }

    val messageRes: Int
        get() = when (this) {
            is ChapterUnavailable -> R.string.err_chapter_unavailable_message
            is UnknownBook -> R.string.err_unknown_book_message
            Offline -> R.string.err_offline_message
            Unexpected -> R.string.err_unexpected_message
        }

    /** Argument for the title/message format, when there is one. */
    val argument: String?
        get() = when (this) {
            is ChapterUnavailable -> reference.formatted
            is UnknownBook -> bookId
            Offline, Unexpected -> null
        }

    companion object {
        fun from(error: Throwable): ReaderError = when (error) {
            is BibleClientException.ContentUnavailable -> ChapterUnavailable(error.reference)
            is BibleClientException.VerseOutOfRange -> ChapterUnavailable(PassageReference(error.reference.bookId, error.reference.chapter))
            is BibleClientException.UnknownBook -> UnknownBook(error.bookId)
            is BibleClientException.NetworkUnavailable -> Offline
            else -> Unexpected
        }
    }
}

File: VerbumKit/Sources/Features/Scripture/ReaderError.swift
import Clients
import Models

/// What the reader can tell the user when a chapter fails to load (spec §52:
/// distinguish causes, never show raw errors).
public enum ReaderError: Error, Equatable, Sendable {
    case chapterUnavailable(PassageReference)
    case unknownBook(BookID)
    case offline
    case unexpected

    init(_ error: any Error) {
        switch error {
        case BibleClientError.contentUnavailable(let reference):
            self = .chapterUnavailable(reference)
        case BibleClientError.verseOutOfRange(let reference, _):
            self = .chapterUnavailable(PassageReference(bookId: reference.bookId, chapter: reference.chapter))
        case BibleClientError.unknownBook(let bookId):
            self = .unknownBook(bookId)
        case BibleClientError.networkUnavailable:
            self = .offline
        default:
            self = .unexpected
        }
    }

    public var title: String {
        switch self {
        case .chapterUnavailable(let reference): L10n.t("\(reference.formatted) isn't available yet")
        case .unknownBook: L10n.t("Unknown book")
        case .offline: L10n.t("You're offline")
        case .unexpected: L10n.t("Couldn't load this chapter")
        }
    }

    public var message: String {
        switch self {
        case .chapterUnavailable:
            L10n.t("This chapter isn't in this translation.")
        case .unknownBook(let bookId):
            L10n.t("“\(bookId)” isn't a book we know.")
        case .offline:
            L10n.t("Connect to load this chapter. Chapters you've already read stay available.")
        case .unexpected:
            L10n.t("Something went wrong on our side. Try again in a moment.")
        }
    }
}
